Feat/add did resolver package

Exulansis opened the pull request:
## Description

Hi everyone, as part of the following accepted [Stacks grant application](https://github.com/stacksgov/Stacks-Grants/issues/61), Jolocom has defined and  developed a new version of the `did:stack:` DID method, based on the new (smart contract based) iteration of the BNS naming layer. This PR contributes both the new DID Method specification, as well as the implementation of the corresponding DID Resolver module.

### A bit of context

The various development milestones were first outlined in [this document](https://www.notion.so/jolocom/Development-Milestones-ca3cc8622b0c4634be5d41b7200af7e3) (you can notice, contributing the developed code to the [Stacks.JS](https://github.com/blockstack/stacks.js) repository is part of the second milestone).
So far the development of this package has taken place in [this repository](https://github.com/jolocom/Stacks-DID-Resolver) hosted by Jolocom. As the main developments were being wrapped up, the migration process (leading to this PR) has been started.

The following [Github issue](https://github.com/jolocom/stacks-did-resolver/issues/17) on the aforementioned repository summarizes both the undertaken developments and remaining open tasks related to the various milestones.

### What is being contributed?

- A DID Method specification document defining the new iteration / version of the Stacks DID method. Can be found [here](https://github.com/jolocom/stacks.js/blob/feat/add_did_resolver_package/packages/did-resolver/docs/DID_Method_Spec.md). Feedback and suggestions for improvement are of course very welcome!
- A new lerna `package` containing the implementation of the corresponding DID resolver module. The source code, including some documentation  and a set of unit / e2e tests can be found [here](https://github.com/jolocom/stacks.js/tree/feat/add_did_resolver_package/packages/did-resolver).
- A [Universal Resolver](https://github.com/decentralized-identity/universal-resolver) driver for `did:stack:v2` identifiers. The corresponding Docker image can be [found here](https://github.com/jolocom/stacks.js/tree/feat/add_did_resolver_package/packages/did-resolver/uni-resolver-driver). *For context, a similar driver for the previous version of the Blockstack DID method can [be found here](https://github.com/decentralized-identity/uni-resolver-driver-did-stack)*.

The contributed package is self contained, no modifications are required to any other included packages / the build system / project setup.

In case any additional information should be submitted as part of this pull request to simplify / aid the review process, please let me know and I can make the requested changes.

Thank you, and looking forward to the emerging conversation!

zone117x committed:
chore: fix #983 and bugs with Clarity int/uint value encoding

zone117x committed:
chore: simplify intCV parsing for unpadded byte array inputs -- bn.js will take care of it as long as we operate on 8-bit boundaries

zone117x committed:
chore: add a few more test vectors for intCV test

zone117x committed:
chore: allow odd-length hex strings that certain libs output

zone117x committed:
fix: bn.js lib accepts strings containing non-integer values and results in weird behavior

zone117x committed:
chore(ci): npm node_modules cache potentially creating flaky CI failures

ahsan-javaid committed:
fix: add missing stx burn and NFT post conditions

friedger committed:
fix: handle empty list in getCVTypeString (#1033)

friedger committed:
fix: use stacks.js repo url

ahsan-javaid committed:
feature: add reward methods for set address in stacking package

friedger committed:
fix: prettify

kyranjamie committed:
ci: improve ci jobs

asimm241 committed:
BREAKING CHANGE: make coreApiUrl readonly for stacks network and initialise in constrctor

asimm241 committed:
fix: removeread only and add private field to prevent run time assignment

asimm241 committed:
fix: fixed lint issue

ahsan-javaid committed:
fix: The transaction ABI validation should accept lists that are less than or equal to the max size specified in the arguments type

kyranjamie committed:
fix: broken types following diverged network config changes

aulneau committed:
fix: remove console.log

aulneau committed:
fix: remove unused const

ahsan-javaid committed:
fix: verify that the public key is a secp256k1 point

zone117x committed:
chore: rebase fixes

Exulansis committed:
Migrated the did-resolver module

Including it as a new package. Currently an error occurs when running
npm run build from the project root. The @stacks/profile dependency is
not resolved, although it is present in node_modules. Needs to be
investigated further

Exulansis committed:
fix(build): Added reference to @stacks/profile

building from the project root works as expected now

Exulansis committed:
feat(v1.0.0-beta): Migrated latest changes

Includes a number of improvements over the previous version, namely:
- Unit tests added in addition to the previously defined integration
  tests
- In-line documentation / comments expanded to cover most relevant
  functions
- Further refactoring and modularisation, removed a number of unused /
  complex functions
- Resolution of migrated BNS names now happens via a helper clarity
  contract
- The DID Method specification has been updated to describe migrated
  DIDs in more detail, in addition to minor editorial changes

Exulansis committed:
fix(package.json): updated metada

Udated the version, license, webpage, repository, and further properties
in the package.json file.

Updated the broken local links in the DID method spec MD file

Exulansis committed:
Commented out service endpoint code

As of now the code for adding the relevant service entries has been
commented out until it's clear which endpoint URL should be included.

Exulansis committed:
docs(README): Expanded section on running tests

Mentioning both the included unit and integration tests.

Exulansis committed:
docs(uni-resolver-driver): Added README

Exulansis committed:
fix(utils): Not using Either in exported helpers

Converted the exported helper functions to return the expected value /
throw an error instead of returning an Either.

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+
+# Abstract
+
+Blockstack is a network for decentralized applications where users own their
```
"Blockstack" should be replace with "Stacks"

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+
+# 1. System Overview
+
+The Blockstack DID method is specified as part of its [decentralized naming
```
Same for "Stacks DID"

Exulansis replied:
Addressed, as per the comment left above.

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+- Off-chain mainnet: `SHB53GD600EMEM74DFMA0B61JN8D8C4VE6DHF2QF`
+- Off-chain testnet: `SJB53GD600EMEM74DFMA0B61JN8D8C4VE4M8NJRP`
+
+The version bytes for the various address types are also documented
```
I am confused about the version code for addresses. Does it matter whether the address of a DID is owned by a multisig wallet? For mainnet, there are now the following version:
*  SP.. (on-chain names and single sig address), 
* SH.. (off-chain names)
* SM.. (multi sig address)
Could you clarify?

friedger commented on packages/did-resolver/README.md:
```diff
@@ -0,0 +1,76 @@
+# @stacks/did-resolver
+
+This repository includes the implementation for the Stack v2 DID resolver, as well as the corresponding `did:stack:v2` [DID method specification](./docs/DID_Method_Spec.md). As outlined in the specification, this module relies on an instance of the [BNS](https://docs.stacks.co/build-apps/references/bns) smart contract, as well as the Atlas peer network to perform DID resolution.
```
Are the blockchain attachments still called "Atlas"?

Exulansis replied:
This is a good question, I would be glad to update all references to Atlas / Gaia / etc. to be consistent across the document. I am not sure what the right formulation would be. I was able to find some sparse documentation on Atlas, but still lack the full picture.

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+subdomains](https://docs.stacks.co/build-apps/references/bns#subdomains), the
+records for which are stored entirely off-chain, but are collectively anchored
+to the blockchain. The ownership and state for these names lives within the P2P
+network data and can be retrieved using the Atlas peer network.
```
Some question for "Atlas"

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+[`name-register`](https://github.com/blockstack/stacks-blockchain/blob/master/src/chainstate/stacks/boot/bns.clar#L609)
+transaction which registered the name.
+
+Blockstack supplies a [reference
```
Stacks supplies..

Exulansis replied:
Addressed :+1:

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+Sample code to generate these TXT records can be found in the [Blockstack Core
+reference implementation](https://github.com/blockstack/blockstack-core), under
+the `blockstack.lib.subdomains` package.  For example, the Python 2 program
+here generates such a TXT record:
```
That library is deprecated and should not be referenced.

Exulansis replied:
Updated, not marking as resolved yet because I'm pretty sure the python snippit included below will need to be updated as well.

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+intermediate middleboxes like CDNs and firewalls can cache data and log
+requests.
+
+# 6. Reference Implementations
```
This looks dated.

Exulansis replied:
The links in the reference implementation section have been updated now. The exact link for the `HTTP API` endpoint for DID resolution will need to be updated shortly (as the appropriate PR adding the endpoint is opened on the [Stacks Blockchain api repository](https://github.com/blockstack/stacks-blockchain-api)).

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+implementation](https://github.com/blockstack/blockstack-core) is available
+under the terms of the General Public Licence, version 3.
+
+# 7. Resources
```
This as well.

Exulansis replied:
Will update this shortly, are there any other places developers should be pointed towards besides the Discord server (at https://discord.gg/YWa8BmKqvR)

friedger commented on packages/did-resolver/src/constants.ts:
```diff
@@ -0,0 +1,78 @@
+
+export const GENESIS_RESOLVER_ADDRESSES = {
+  main: 'SPPJE6KB9CNCVTS9RAFHY1RSXAH381W8RKJDM9J9.BNS-migrated-helper',
+  test: 'STPJE6KB9CNCVTS9RAFHY1RSXAH381W8RJQV1YCB.BNS-migrated-helper',
```
That contract does not exist since the reset of the testnet

Exulansis replied:
I just noticed the network was reset. I assume the reset happens regularly, and it makes no sense to reference a specific address for the testnet deployment then, is this correct?

friedger commented on packages/did-resolver/docs/DID_Method_Spec.md:
```diff
@@ -0,0 +1,769 @@
+and [3.4](#34-deactivating-a-stacks-did) outline the respective interactions in
+more detail.
+
+A resolvable Stacks DID can be derived for any on-chain name by concatenating
```
What about names owned by a contract? Should these names be excluded? Or should the documentation and lib updated to for names owned by contract principals?

friedger reviewed (changes requested):
Great work!

I have reviewed mainly the documentation, not the code.

I have questions about names owned by multi-sig addresses and contracts.

Exulansis committed:
docs(DID method): Replaced Blockstack with Stacks

In accordance with the received feedback, references to Blockstack / the
Blockstack DID method / Blockstack blockchain / etc. were replaced with references to Stacks.

Also restored the formatting to the code snippits, they accidently got
mangled during the previous commit.

kyranjamie commented on packages/did-resolver/src/api.ts:
```diff
@@ -0,0 +1,203 @@
+ * further useful properties
+ */
+
+export const fetchNameInfo = (
+  network: StacksNetwork,
+  {
+    name,
+    namespace,
+  }: {
+    name: string
+    namespace: string
+  }
+): FutureInstance<Error, NameInfo> => {
```
This seems like an unusual API choice to me, have you considered either:

1) **Object argument**. The function takes a single object argument including the network
2) **Currying**. Signature `(network: StacksNetwork) => ({ name, namespace }) => FutureInstance<Error, NameInfo>`

The latter allows reuse of the resulting function without repetition of the network arg.

Exulansis replied:
Thank you for the remark, will update this call (and other similar occurrences) to use an object argument. I believe I originally defined these functions to make use of currying, but the partially applied versions of the functions did not end up being too useful given the current structure, so I simplified them but did not change the way the arguments are passed.
